### PR TITLE
Simple  list comprehensions.

### PR DESCRIPTION
**Pull request by Matthiaas**
This adds very simple list comprehensions to ocaml.

### CODE REVIEW

File: lambda/translcore.ml [hunk 1]
```diff
@@ -152,16 +152,16 @@ let rec push_defaults loc bindings cases partial =
   | _ ->
       cases
 
-let comprehension_prim ~is_array ~dir= 
-  let prefix = if is_array then "array_" else "" in
+let comprehension_prim ~is_concat ~dir= 
+  let prefix = if is_concat then "concat_" else "" in
   let fun_name =  prefix ^
```
Comment (lpw25) [inline lambda/translcore.ml]: This is a bit awkward to read. Better to name the suffix too and then do:
```ocaml
let fun_name = prefix ^ suffix in
```

File: typing/typedtree.mli [hunk 2]
```diff
@@ -248,6 +248,10 @@ and expression_desc =
   | Texp_ifthenelse of expression * expression * expression option
   | Texp_sequence of expression * expression
   | Texp_while of expression * expression
+  | Texp_list_comprehension of 
+      Ident.t list list * expression * comprehension list list
```
Comment (lpw25) [inline typing/typedtree.mli]: I think it would be better to include the `Ident.t`s in the `comprehension` type.

File: typing/typecore.mli [hunk 3]
```diff
@@ -31,6 +31,7 @@ type type_forcing_context =
   | If_no_else_branch
   | While_loop_conditional
   | While_loop_body
+  | InComprehensionArgument
```
Comment (lpw25) [inline typing/typecore.mli]: This should be `In_comprehension_argument` because OCaml traditionally uses snake case.

File: parsing/parser.mly [hunk 4]
```diff
@@ -2309,6 +2309,24 @@ simple_expr:
...
       { unclosed "(" $loc($1) ")" $loc($6) }
 ;
+
+inner_comprehension:
+|  ext_attributes pattern EQUAL simple_expr direction_flag simple_expr
+      { From_to($2, $4, $6, $5) }
+|  ext_attributes pattern IN simple_expr { In($2, $4) }
+;
+
+// List are reveresed s.t. we can Left fold over it in the typecore.
```
Comment (lpw25) [inline parsing/parser.mly]: Since these lists will not be long, it's probably better to store them in the AST in the order that someone would expect rather than reversed.

File: typing/typecore.ml [hunk 5]
```diff
@@ -5089,6 +5139,49 @@ and type_andops env sarg sands expected_ty =
   let let_arg, rev_ands = loop env sarg (List.rev sands) expected_ty in
   let_arg, List.rev rev_ands
 
+(*TODO mbungeroth: Why do I need the type annotation here to make this work?*)
+  and comprehension ~loc ~env ~ty_expected ~container_type 
+                  ~sbody ~(comp_typell : Parsetree.comprehension list list) = 
+
+  let from_to_comprehension ~body_env ~env param slow shigh dir =
```
Comment (lpw25) [inline typing/typecore.ml]: `from_to_comprehension` and `in_comprehension` can probably just be inlined into `call`. I'd also probably lift `call` out of `comprehension` and call it something like `type_comprehension_clause`.

**Review by lpw25 — COMMENTED** [review]
I've looked over some of the code. I still need to read typecore.ml and translcore.ml in detail.

File: typing/typecore.ml [hunk 6]
```diff
@@ -595,6 +596,23 @@ let split_cases env cases =
                     Mixed_value_and_exception_patterns_under_guard))
     | vp, ep -> add_case vals case vp, add_case exns case ep
   ) cases ([], [])
+ 
+let type_for_loop ~loc ~env ~param ty=
```
Comment (lpw25) [inline typing/typecore.ml]: Should be `type_for_loop_index` since it only "types" the index of the for loop.

File: typing/typecore.ml [hunk 7]
```diff
@@ -3068,26 +3089,55 @@ and type_expect_
...
+  | Pexp_list_comprehension (sbody, comp_typell) ->  
+    if !Clflags.principal then begin_def ();
+    let without_list_ty = Ctype.newvar ()  in
+    unify_exp_types loc env 
+      (instance (Predef.type_list without_list_ty)) (instance ty_expected);
+    if !Clflags.principal then begin 
+      end_def();
+      generalize_structure without_list_ty; 
+    end;
+
```
Comment (lpw25) [inline typing/typecore.ml]: Normally OCaml functions should not contain blank lines.

File: typing/typecore.ml (hunk 5 above)
Comment (Matthiaas) [reply typing/typecore.ml]: Why should I pull type_comprehension_clause out if its only used in comprehension?

File: typing/typecore.ml [hunk 8]
```diff
@@ -5089,6 +5139,49 @@ and type_andops env sarg sands expected_ty =
...
+            List.fold_left 
+                (fun acc comp_type -> call acc ~env ~comp_type ) 
+                ([], [], env)  
+                comp_typel 
+            in
+            new_ids::ids, new_comps::comps, new_env
+    ) 
+    ([], [], env) comp_typell
+  in
+  let body = type_expect new_env sbody ty_expected in
```
Comment (lpw25) [inline typing/typecore.ml]: I'd be tempted to split the typing of the body out of this function, and have this one return `ids, env, comps`. That way you can avoid needing to pass in `ty_expected` and `sbody`.

File: typing/typecore.ml [hunk 9]
```diff
@@ -5089,6 +5139,49 @@ and type_andops env sarg sands expected_ty =
...
+    (*let body = type_expect new_env sbody ty_expected in*)
+    id, From_to(param, low, high, dir), new_env
+  in
+  let in_comprehension ~body_env ~env param siter= 
+    let item_ty = newvar() in
+    let iter_ty = instance (container_type item_ty) in
+    let iter = type_expect env siter 
+        (mk_expected ~explanation:InComprehensionArgument iter_ty) in
+    let id, new_env = type_for_loop ~loc ~env:body_env ~param item_ty
+    in
```
Comment (lpw25) [inline typing/typecore.ml:5142]: `in` should be on the previous line here. The two standard ways to indent a `let` are:
```ocaml
let ... = ... in
```
and
```ocaml
let ... =
  ....
in
```
You will see some variations of that in the compiler, but its still best to just use those two.

File: lambda/translcore.ml [hunk 10]
```diff
@@ -447,6 +459,75 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
+      | From_to (_,e2,e3, dir) -> 
+          Pintval, [transl_exp ~scopes e2; transl_exp ~scopes e3],
+          comprehension_prim ~is_concat ~dir
+      | In (_, e2) ->  
+          Pgenval, [transl_exp ~scopes e2],
+          list_comp_in_prim ~is_concat
+      in
+      let fn = Lfunction {kind = Curried;
+                          params= [param, pval];
+                          return = return;
```
Comment (lpw25) [inline lambda/translcore.ml]: I think using `return` here is only correct for the case where `bdy` equals `body`. I think in the other cases the functions instead return a list.

**Review by lpw25 — COMMENTED** [review]
Finshed going through the rest

File: typing/typecore.ml (hunk 5 above)
Comment (lpw25) [reply typing/typecore.ml]: Good question. I think that once a function gets to a certain size it is better to lift it out. For example, `type_function` is only called from `type_expect`, but having it as a local function in there would make things much harder to read. I also think it is slightly better to lift things out if they make semantic sense on their own.

I think both those points are debatable here, so it's reasonable to do it either way.

File: lambda/translcore.ml (hunk 10 above)
Comment (Matthiaas) [reply lambda/translcore.ml]: Does that mean I should use Pgenval for the case when it is a list?

File: lambda/translcore.ml [hunk 11]
```diff
@@ -447,6 +459,188 @@ and transl_exp0 ~in_new_scope ~scopes e =
   | Texp_while(cond, body) ->
       Lwhile(transl_exp ~scopes cond,
              event_before ~scopes body (transl_exp ~scopes body))
+  
+  | Texp_arr_comprehension (body, blocks) ->
```
Comment (Matthiaas) [inline lambda/translcore.ml]: This is till a little cluttery. I should probably move some functions out. Any suggestions?

File: lambda/translcore.ml (hunk 10 above)
Comment (lpw25) [reply lambda/translcore.ml]: Yeah, Pgenval would be right for a list.

File: parsing/ast_mapper.ml [hunk 12]
```diff
@@ -417,22 +417,30 @@ module E = struct
...
-        list_comprehension ~loc ~attrs (sub.expr sub e1) comp
-       *)
-    (*TODO mbungeroth: Add usefull code here.*)
-    assert false
-    | Pexp_arr_comprehension _-> assert false
+    | Pexp_list_comprehension (e1, comp_types) -> 
+      list_comprehension ~loc ~attrs (sub.expr sub e1)
+        (List.map (fun comp_types -> 
+          List.map (fun comp_type ->
+            match comp_type with
```
Comment (lpw25) [inline parsing/ast_mapper.ml]: You should probably lift this part out as `map_comprehension` and share it with the array case.

File: parsing/ast_iterator.ml [hunk 13]
```diff
@@ -379,18 +379,19 @@ module E = struct
...
-      TODO mbungeroth: to this recursivly
-      (match comp_type with
-      | From_to (p,e2,e3,_) ->
-        sub.expr sub e1; sub.pat sub p; sub.expr sub e2;
+    | Pexp_list_comprehension (e1, comp_types)
+    | Pexp_arr_comprehension (e1, comp_types) ->
+      sub.expr sub e1; 
+      (List.iter (fun comp_types -> 
+        List.iter (fun comp_type ->
+          match comp_type with
```
Comment (lpw25) [inline parsing/ast_iterator.ml]: Might as well lift his one out as `iter_comprehension` too.

File: typing/typecore.ml [hunk 14]
```diff
@@ -5139,48 +5136,49 @@ and type_andops env sarg sands expected_ty =
...
-            new_comps::comps, new_env
-    ) 
-    comp_typell ([], env) 
-  in
-  let body = type_expect new_env sbody ty_expected in
-  body, comps
+(*TODO mbungeroth: Why do I need the type annotation here to make this work?*)
+  and type_comprehension_clause ~loc ~env ~container_type 
+      ~(comp_typell : Parsetree.comprehension list list) = 
+    let type_comprehension (comps, body_env) ~env
```
Comment (lpw25) [inline typing/typecore.ml]: List this out, remove the accumulation machinery and inline the two sub-functions into it.

File: lambda/translcore.ml [hunk 15]
```diff
@@ -460,99 +460,169 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
-            let loop = 
-              Lfor(index, 
-                (int 0), (l_int_op Psubint len (int 1)), Upto, 
-                (Llet(
-                  Strict,Pintval, id, (l_array_get  in_ (Lvar(index))), 
-                  inner_body)))
-            in 
-            loop, l_int_op Pmulint inner_len len
-            
+        let in_ = (transl_exp ~scopes e2) in
```
Comment (lpw25) [inline lambda/translcore.ml]: Unnecessary parens

File: lambda/translcore.ml [hunk 16]
```diff
@@ -460,99 +460,169 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
+        let len = Lprim( (Parraylength(Pgenarray)), [in_], Loc_unknown) in
```
Comment (lpw25) [inline lambda/translcore.ml]: This length expression is going to appear twice: once in the bounds of the loop and once when computing the total length. It won't make much difference -- since the read should be in cache -- but I think it's reasonably easy to avoid the repetition.

Rather than using the `len` expression directly in bounds of the loop, have it use a fresh variable and have this function return the variable along with the length expression. Then change `transl_loops` to return a list of pairs of length variables and length expressions instead of a single combined length expression. Finally, you can generate `let` bindings around the outside of the loop that bind the lengths to the length variables so that they can be shared between both uses.

File: lambda/translcore.ml [hunk 17]
```diff
@@ -460,99 +460,169 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
+        Lassign(counter, transl_int_op Paddint (int 1) (Lvar(counter))))
+      in
+      let base_body_with_global = 
+        if is_only_block then
+          (*If there is only one block then globalcounter equals counter.*) 
+          base_body
+        else 
+          Lsequence( 
+            base_body,
+            Lassign(
```
Comment (lpw25) [inline lambda/translcore.ml]: Rather than adding 1 each iteration of the inner loop, I think you can just add `res_length` or `counter`.

File: lambda/translcore.ml (hunk 11 above)
Comment (lpw25) [reply lambda/translcore.ml]: I think you should at least lift this whole piece out as its own `transl_arr_comprehension` function.

File: lambda/translcore.ml [hunk 18]
```diff
@@ -460,99 +460,169 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
+      let base_loop = transl_loop ~type_comp:(List.hd block) ~body:base_body in
+      List.fold_left (fun (body, len) type_comp ->
+        let new_body, new_len = transl_loop ~type_comp ~body in
+        new_body, transl_int_op Pmulint new_len len 
+        ) base_loop (List.tl block)
+    in       
+    (*The block created here is the most inner block. This needs to execute the
+      execute the actual body of the comprehension and write it into the 
+      correct result array.*)   
+    let transl_base_block block gloabl_counter is_only_block =
```
Comment (lpw25) [inline lambda/translcore.ml]: I think that the only real difference between `transl_base_block` and `transl_block` is that `transl_base_block` has to run the first iteration before it makes the array since it might be creating a `Pgenarray` (i.e. array that may or may not hold floats) whilst `transl_block` knows it is creating a `Paddrarray` (i.e. an array that is definitely doesn't hold floats).

I think that it would be slightly better to have a single recursive `transl_blocks` function that took an `array_kind` as a parameter, saying what kind of array was being created. If the kind is `Pgenarray` then it does the extra work to run the first iteration before making the array, otherwise it just does the simple version. When this function recursives to handle nested blocks it would pass in `Paddrarray` since it would know that the nested arrays always held pointers to arrays.

File: lambda/translcore.ml [hunk 19]
```diff
@@ -460,99 +460,169 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
+          let sub_arr = Ident.create_local "sub_arr" in 
+          Lfor(i, (int 0), (transl_int_op Psubint len (int 1)), Upto, 
+            Llet(Strict, Pgenval, sub_arr, transl_array_get arr (Lvar(i)),
+              transl_for (Lvar(sub_arr)) (block_count - 1)))
+      in 
+      Llet(Strict, Pgenval, res, empty_arr,
+        Llet(Variable, Pintval, counter, (int 0),  
+          Lsequence(transl_for arr block_count, Lvar(res))))
+    in
+    let gloabl_counter = Ident.create_local "gloabl_counter" in
```
Comment (lpw25) [inline lambda/translcore.ml]: s/gloabl/global/

**Review by lpw25 — COMMENTED** [review]
Some comments on commit 6b7d832

File: lambda/translcore.ml [hunk 20]
```diff
@@ -592,51 +603,91 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
-        then Pgenval 
-        else Typeopt.value_kind body.exp_env body.exp_type 
+    let int n = Lconst (Const_base (Const_int n)) in  
+    let block = List.fold_left (fun acc block -> 
+        let rec loop = function 
+          | [] -> assert false 
+          | [last] -> [last, block.guard]
+          | l::ls -> (l, None) :: (loop ls) 
+        in
+        List.concat [List.rev (loop block.clauses); acc]
```
Comment (lpw25) [inline lambda/translcore.ml]: This is `List.rev_append`

File: lambda/translcore.ml [hunk 21]
```diff
@@ -592,51 +603,91 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
+          let bdy = 
+            Lprim( 
+              (*TODO: change None to usefull P...val*)
+              Pmakeblock(0, Immutable, None), 
+              [(transl_exp ~scopes  body); Lvar(res)], Loc_unknown) 
+          in
+          false, last, bdy, Typeopt.value_kind body.exp_env body.exp_type
+        | el::ls ->  
+          true, el, (transl_list_comp ls (Some (Lvar(res)))), Pgenval
+        | _ -> assert false
```
Comment (lpw25) [inline lambda/translcore.ml]: Much better to use `[]` rather than `_` here.

File: lambda/translcore.ml [hunk 22]
```diff
@@ -152,18 +152,23 @@ let rec push_defaults loc bindings cases partial =
...
+  let suffix = if single_loop then "" else "_prepend" in
+  let function_name = match dir with 
     | Upto ->  "map_from_to"
     | Downto -> "map_from_downto"  in 
-  let fun_name =  prefix ^ suffix in
+  let fun_name =  String.concat "" [prefix;function_name;suffix] in
   Lambda.transl_prim "CamlinternalComprehension" fun_name
 
-let list_comp_in_prim ~is_concat = 
+let in_from_to_comp_prim ~is_concat ~single_loop =
```
Comment (lpw25) [inline lambda/translcore.ml]: `in` and `from_to` is a bit confusing. Isn't it only used for the `in` cases?

File: parsing/ast_mapper.ml [hunk 23]
```diff
@@ -419,27 +419,33 @@ module E = struct
...
-          List.map (fun comp_type ->
-            match comp_type with
-            | From_to (p, e2, e3, dir) ->
-              From_to(sub.pat sub p, sub.expr sub e2,
-            sub.expr sub e3, dir)
-            | In (p, e2) -> 
-            In(sub.pat sub p, sub.expr sub e2)
-          )comp_types 
+        (List.map (fun {clauses=comp_types; guard}  -> 
+          let clauses =
```
Comment (lpw25) [inline parsing/ast_mapper.ml]: Better to lift this out as a `map_comprehension` function. Keeping the cases in these functions, by lifting out the mapping of sub-components, makes them very regular and much easier to review for correctness.

File: stdlib/camlinternalComprehension.ml [hunk 24]
```diff
@@ -1,11 +1,17 @@
 
-let map = List.map
+let map f y = 
+  let rec loop f y res =
```
Comment (lpw25) [inline stdlib/camlinternalComprehension.ml]: This is `List.rev_map`

File: stdlib/camlinternalComprehension.ml [hunk 25]
```diff
@@ -1,11 +1,17 @@
 
-let map = List.map
+let map f y =
```
Comment (lpw25) [inline stdlib/camlinternalComprehension.ml]: I think it is reasonable to change to the tail-recursive map. However, I think that means that there isn't any need to treat the single loop case specially, since it is going to be built and then reversed either way.

File: stdlib/camlinternalComprehension.ml [hunk 26]
```diff
@@ -14,23 +20,57 @@ let map_from_downto f from to_ =
...
 *)
-let concat_map f l = 
-  List.concat (List.map f l)
-;;
 
-let concat_map_from_to f from to_ = 
-  List.concat (map_from_to f from to_ )
+let rev = List.rev;;
+
+let concat_map_prepend f l res=
```
Comment (lpw25) [inline stdlib/camlinternalComprehension.ml]: I think the concat versions are the same as the non-concat versions.

File: lambda/translcore.ml [hunk 27]
```diff
@@ -592,51 +603,91 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
+    let block = List.fold_left (fun acc block -> 
+        let rec loop = function 
+          | [] -> assert false 
+          | [last] -> [last, block.guard]
+          | l::ls -> (l, None) :: (loop ls) 
+        in
+        List.concat [List.rev (loop block.clauses); acc]
+        ) [] blocks
+    in
+    let rec transl_list_comp block upper_res=
```
Comment (lpw25) [inline lambda/translcore.ml]: This function is very messy and hard to follow. I think that most of the complexity will disappear by the observation that there is no need for separate `concat` and `prepend` versions of the primitives, or any need for different handling of the `single_loop` case. However, I think it would be useful to think about and discuss how you might implement this function more clearly.

**Review by lpw25 — COMMENTED** [review]
Some comments on the latest commit.

File: stdlib/camlinternalComprehension.ml (hunk 24 above)
Comment (lpw25) [reply stdlib/camlinternalComprehension.ml]: Actually I misread the code. I think it's actually `List.fold_left`.

File: lambda/translcore.ml (hunk 22 above)
Comment (Matthiaas) [reply lambda/translcore.ml]: Yes.

File: lambda/translcore.ml (hunk 16 above)
Comment (Matthiaas) [reply lambda/translcore.ml]: I dont know exactly how Ocaml does that but should a length lookup be the same as a variable lookup?
So this would even add overhead but having another variable.

Comment (lpw25) [reply lambda/translcore.ml]: The length lookup will be a read from memory, whilst reading a mutable variable will be using a register. Sometimes it is better to re-read a value from memory -- e.g. if you are going to hold the value live for a long time -- but here I expect using a variable to be better.

Comment (Matthiaas) [reply lambda/translcore.ml]: Sadly, its not as easy as just defining  a new Let variable in this part of the code. Since then it will be out of scope when used the second time. 
I think transl_loops would need to return a (Ident.t * lambda) list s.t the caller can the init all the lengths such that they are in the correct scope. 
Any suggestions?

Comment (lpw25) [reply lambda/translcore.ml]: Having `transl_loops` return `(Ident.t * lambda) list` is what I was suggesting above. I think it is probably the best way to write this code.

Comment (Matthiaas) [reply lambda/translcore.ml]: Oh, yeah. Thanks!

File: parsing/ast_iterator.ml [hunk 28]
```diff
@@ -342,6 +342,19 @@ module E = struct
   let iter sub {pexp_loc = loc; pexp_desc = desc; pexp_attributes = attrs} =
     sub.location sub loc;
     sub.attributes sub attrs;
+    let iter_comprehension comp_types=
```
Comment (lpw25) [inline parsing/ast_iterator.ml]: It would be a little more consistent to lift this function outside of `iter`. It would also avoid cluttering up the main `iter` function.

File: parsing/ast_mapper.ml [hunk 29]
```diff
@@ -381,6 +381,21 @@ module E = struct
     let open Exp in
     let loc = sub.location sub loc in
     let attrs = sub.attributes sub attrs in
+    let map_comprehension comp_types=
```
Comment (lpw25) [inline parsing/ast_mapper.ml]: Slightly better and more consistent to lift this out of `map`.

File: lambda/translcore.ml [hunk 30]
```diff
@@ -467,247 +471,73 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
-        | None -> translated_blocks
-        | Some gc -> 
-          Llet(
-            Variable, Pintval, gc, int 0,
-            Llet(
-              Strict, Pgenval, arr,
-              translated_blocks,
-              transl_concat_arrays (Lvar(arr)) block_count gc
-            )))
+    let loc = (of_location ~scopes e.exp_loc) in
```
Comment (lpw25) [inline lambda/translcore.ml]: Unnecessary parens

File: lambda/translcore.ml [hunk 31]
```diff
@@ -467,247 +471,73 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
-              from_to_comp_prim ~is_concat ~single_loop ~dir
-        | In (param, _, e2) ->  
-            param, Pgenval, (transl_exp ~scopes e2)::args,
-              in_from_to_comp_prim ~is_concat ~single_loop
+            [transl_exp ~scopes e2; transl_exp ~scopes e3; Lvar(new_acc)],
+            from_to_comp_prim ~dir,
+            body
+        | In (pat, e2) ->  
+            let pat_id = Ident.create_local "pat" in
+            pat_id , Pgenval,
```
Comment (lpw25) [inline lambda/translcore.ml]: This `Pgenval` can be replaced with `Typeopt.value_kind pat.pat_env pat.pat_type`.

File: lambda/translcore.ml [hunk 32]
```diff
@@ -1276,6 +1106,284 @@ and transl_record ~scopes loc env fields repres opt_init_expr =
...
+      let in_kind = Typeopt.array_kind e2 in
+      let len = Lprim( (Parraylength(in_kind)), [in_], loc) in
+      let index = Ident.create_local "index" in    
+      Lfor(index, (int 0), Lprim(Psubint, [Lvar(len_var); int 1], loc) , Upto, 
+        Matching.for_let ~scopes pat.pat_loc
+          (Lprim(Parrayrefs(in_kind), [in_; Lvar(index)], loc)) pat body), 
+      (len_var,len)
+    | From_to(id, _, e2, e3, dir) ->  
+      let from = (transl_exp ~scopes e2) in
+      let to_ = (transl_exp ~scopes e3) in
```
Comment (lpw25) [inline lambda/translcore.ml]: Unnecessary parens

File: lambda/translcore.ml [hunk 33]
```diff
@@ -467,247 +471,73 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
         | From_to (param, _,e2,e3, dir) -> 
             param, Pintval, 
-              (transl_exp ~scopes e2)::(transl_exp ~scopes e3)::args,
-              from_to_comp_prim ~is_concat ~single_loop ~dir
-        | In (param, _, e2) ->  
-            param, Pgenval, (transl_exp ~scopes e2)::args,
-              in_from_to_comp_prim ~is_concat ~single_loop
+            [transl_exp ~scopes e2; transl_exp ~scopes e3; Lvar(new_acc)],
+            from_to_comp_prim ~dir,
+            body
```
Comment (lpw25) [inline lambda/translcore.ml]: These kinds of tuple expressions can be hard to read. I think it is generally better to write them as e.g.:
```ocaml
let pval = Pintval in
let args = [transl_exp ~scopes e2; transl_exp ~scopes e3; Lvar(new_acc)] in
let func = from_to_comp_prim ~dir in
param, pval, args, func, body
```
that way the components get meaningful names that match what they will be bound to be the surrounding let.

File: lambda/translcore.ml [hunk 34]
```diff
@@ -1276,6 +1106,284 @@ and transl_record ~scopes loc env fields repres opt_init_expr =
...
+          Lsequence(transl_for arr arr_len arrs, Lvar(res))))
+    | Pintarray | Pfloatarray | Paddrarray as kind -> 
+      Llet(Strict, Pgenval, res, make_array_of_kind kind (Lvar(total_len)),
+        Llet(Variable, Pintval, counter, (int 0),  
+          Lsequence(transl_for arr arr_len arrs, Lvar(res))))
+  in
+  (*The global counter is required when the size of the result array is not 
+    known in advance.*)
+  let global_counter = 
+    if (List.length blocks  = 1) && Option.is_none (List.hd blocks).guard
```
Comment (lpw25) [inline lambda/translcore.ml]: I think this is clearer as a pattern match:
```ocaml
match blocks with
| [{ guard = None; _}] -> None
| _ -> Some (Ident.create_local "global_counter")
```

File: lambda/translcore.ml [hunk 35]
```diff
@@ -1276,6 +1106,284 @@ and transl_record ~scopes loc env fields repres opt_init_expr =
...
+          transl_concat_arrays (Lvar(arr)) (Lvar(len)) arrs array_kind gc)))
+  | Unguarded(body, _kind), None -> body 
+  | Guarded(body_and_len, _, id), Some gc -> 
+    let body = Ident.create_local "body" in
+    let len = Ident.create_local "len" in
+    let arr = Ident.create_local "arr" in 
+    Llet(
+      Variable, Pintval, gc, int 0,
+      Lstaticcatch(body_and_len, (id, [(body, Pgenval); (len, Pintval)]),
+        Llet(Strict, Pgenval, arr, Lvar(body),
```
Comment (lpw25) [inline lambda/translcore.ml]: No need for this additional let, you can just use `Lvar(body)` instead of `Lvar(arr)` -- although I might use the name `arr` rather than `body` (and `arr_and_len`).

**Review by lpw25 — COMMENTED** [review]
Some comments on 565835a9

File: parsing/ast_iterator.ml (hunk 28 above)
Comment (Matthiaas) [reply parsing/ast_iterator.ml]: Not needed anymore, cause it is an extension now.

File: parsing/ast_mapper.ml (hunk 29 above)
Comment (Matthiaas) [reply parsing/ast_mapper.ml]: Not needed anymore, cause it is an extension now.

**Comment by lpw25** [comment]
Closing since this was merged into the flambda-backend repo directly.
